projects/lights_control/src/lights_gpio.c

```diff
@@ -2,211 +2,113 @@
-
-static LightsBoard s_board;
-
-static const GPIOAddress *s_address_front_lookup[] = {
-  [LIGHTS_EVENT_HORN] = s_horn_peripheral_addresses,                        //
-  [LIGHTS_EVENT_HIGH_BEAMS] = s_high_beams_peripheral_addresses,            //
-  [LIGHTS_EVENT_LOW_BEAMS] = s_low_beams_peripheral_addresses,              //
-  [LIGHTS_EVENT_DRL] = s_drl_peripheral_addresses,                          //
-  [LIGHTS_EVENT_SIGNAL_LEFT] = s_front_signal_left_peripheral_addresses,    //
-  [LIGHTS_EVENT_SIGNAL_RIGHT] = s_front_signal_right_peripheral_addresses,  //
-};
-
-static const GPIOAddress *s_address_rear_lookup[] = {
-  [LIGHTS_EVENT_BRAKES] = s_brakes_peripheral_addresses,
-  [LIGHTS_EVENT_STROBE] = s_strobe_peripheral_addresses,
-  [LIGHTS_EVENT_SIGNAL_LEFT] = s_rear_signal_left_peripheral_addresses,
-  [LIGHTS_EVENT_SIGNAL_RIGHT] = s_rear_signal_right_peripheral_addresses,
-};
-
-static StatusCode prv_init_peripheral(GPIOAddress *peripherals) {
+StatusCode prv_set_peripherals(uint16_t bitset, GPIOState state) {
   uint8_t i;
-
-  // same settings are used for all ports
-  GPIOSettings settings = { .direction = GPIO_DIR_OUT,
-                            .state = GPIO_STATE_HIGH,
-                            .resistor = GPIO_RES_NONE,
-                            .alt_function = GPIO_ALTFN_NONE };
-
-  for (i = 0; i < SIZEOF_ARRAY(peripherals); i++) {
-    StatusCode status = gpio_init_pin(&peripherals[i], &settings);
-    if (status != STATUS_CODE_OK) {
-      return status;
+  while (bitset) {
+    i = __builtin_ffs(bitset) - 1;  // index of first 1 bit
+    StatusCode s = gpio_set_state(&s_addresses[i], state);
+    if (s) {
+      return s;
     }
+    bitset &= ~(1 << i);  // bit is read, so we clear it
   }
   return STATUS_CODE_OK;
 }
 
-StatusCode lights_gpio_init() {
-  uint8_t i;
+static StatusCode prv_get_board_type(LightsBoard *board, LightsConfig *config) {
+  // initializing board type pin
+  StatusCode board_type_init_status =
+      gpio_init_pin(config->board_type_address, config->gpio_settings_in);
+  if (board_type_init_status) {
+    return board_type_init_status;
+  }
+  // reading the state to know the board type
+  GPIOState state;
+  StatusCode state_status = gpio_get_state(config->board_type_address, &state);
+  if (state_status) {
+    return state_status;
+  }
+  *board = (state) ? LIGHTS_BOARD_FRONT : LIGHTS_BOARD_REAR;
+  return STATUS_CODE_OK;
+}
 
-  const GPIOAddress **address_lookup;
-  address_lookup = (s_board == LIGHTS_BOARD_FRONT) ? s_address_front_lookup : s_address_rear_lookup;
+StatusCode lights_gpio_init(LightsGPIOInitMode mode) {
+  LightsConfig *config = lights_config_load();
+  // test modes are used for unit tests
+  if (mode == LIGHTS_GPIO_INIT_MODE_NORMAL) {
+    StatusCode s = prv_get_board_type(&s_board, config);
+    if (s) {
+      return s;
+    }
+  } else {
+    s_board = (mode == LIGHTS_GPIO_INIT_MODE_TEST_FRONT) ? LIGHTS_BOARD_FRONT : LIGHTS_BOARD_REAR;
```

Comment:
What's the point of this? Why not just set the x86 GPIO state?

Reply:
oh ya tru